src/legacy/core_plugins/kibana/public/management/sections/objects/components/object_view/form.tsx

```diff
@@ -0,0 +1,280 @@
+            <EuiFlexItem grow={false}>
+              <EuiButton
+                fill={true}
+                aria-label={i18n.translate('kbn.management.objects.view.saveButtonAriaLabel', {
+                  defaultMessage: 'Save { title } object',
+                  values: {
+                    title: service.type,
+                  },
+                })}
+                onClick={this.onSubmit}
+                disabled={!isValid}
+                data-test-subj="savedObjectEditSave"
+              >
+                <FormattedMessage
+                  id="kbn.management.objects.view.saveButtonLabel"
+                  defaultMessage="Save { title } object"
+                  values={{ title: service.type }}
+                />
+              </EuiButton>
+            </EuiFlexItem>
+          )}
+
+          <EuiFlexItem grow={false}>
+            <EuiButtonEmpty
+              aria-label={i18n.translate('kbn.management.objects.view.cancelButtonAriaLabel', {
+                defaultMessage: 'Cancel',
+              })}
+              onClick={this.onCancel}
+              data-test-subj="savedObjectEditCancel"
+            >
+              <FormattedMessage
+                id="kbn.management.objects.view.cancelButtonLabel"
+                defaultMessage="Cancel"
+              />
+            </EuiButtonEmpty>
+          </EuiFlexItem>
+        </EuiFlexGroup>
+      </EuiForm>
+    );
+  }
+
+  handleFieldChange = (name: string, newState: FieldState) => {
+    this.setState({
+      fieldStates: {
+        ...this.state.fieldStates,
+        [name]: newState,
+      },
+    });
+  };
+
+  isFormValid() {
+    const { fieldStates } = this.state;
+    return !Object.values(fieldStates).some(state => state.invalid === true);
+  }
+
+  onCancel = () => {
+    window.history.back();
+  };
+
+  onSubmit = async () => {
+    const { object, onSave } = this.props;
+    const { fields, fieldStates } = this.state;
+
+    if (!this.isFormValid()) {
+      return;
+    }
+
+    const source = cloneDeep(object.attributes as any);
+    fields.forEach(field => {
+      let value = fieldStates[field.name]?.value ?? field.value;
+
+      if (field.type === 'array' && typeof value === 'string') {
+        value = JSON.parse(value);
+      }
+
+      set(source, field.name, value);
+    });
+
+    const { references, ...attributes } = source;
+
+    onSave({ attributes, references });
+  };
+}
+
+/**
+ * Creates a field definition and pushes it to the memo stack. This function
+ * is designed to be used in conjunction with _.reduce(). If the
+ * values is plain object it will recurse through all the keys till it hits
+ * a string, number or an array.
+ *
+ * @param {string} key The key of the field
+ * @param {mixed} value The value of the field
+ * @param {array} parents The parent keys to the field
+ * @returns {array}
+ */
+const recursiveCreateFields = (key: string, value: any, parents: string[] = []): ObjectField[] => {
```

Comment:
Should this have a max depth safety mechanism to avoid recursing deeply nested objects if one were to exist?

Reply:
Doesn't cost much to add. Also extracted the field generation logic and added tests